internal/messagequeue/messagequeue.go

```diff
@@ -539,74 +545,74 @@ func (mq *MessageQueue) pendingWorkCount() int {
 
 // Convert the lists of wants into a Bitswap message
 func (mq *MessageQueue) extractOutgoingMessage(supportsHave bool) bsmsg.BitSwapMessage {
-	mq.wllock.Lock()
-	defer mq.wllock.Unlock()
-
 	// Get broadcast and regular wantlist entries.
-	// SortedEntries() slows down the MessageQueue a lot, and entries only need
-	// to be sorted if the number of wants will overflow the size of the
-	// message (to make sure that the highest priority wants are sent in the
-	// first message).
-	// We prioritize cancels, then regular wants, then broadcast wants.
-	var peerEntries []bswl.Entry
-	var bcstEntries []bswl.Entry
-	maxCancelsSize := mq.cancels.Len() * bsmsg.MaxEntrySize
-	maxPeerSize := mq.peerWants.pending.Len() * bsmsg.MaxEntrySize
-	maxBcstSize := mq.bcstWants.pending.Len() * bsmsg.MaxEntrySize
-
-	if maxCancelsSize+maxPeerSize < mq.maxMessageSize {
-		peerEntries = mq.peerWants.pending.Entries()
-	} else {
-		peerEntries = mq.peerWants.pending.SortedEntries()
-	}
-	if maxCancelsSize+maxPeerSize+maxBcstSize < mq.maxMessageSize {
-		bcstEntries = mq.bcstWants.pending.Entries()
-	} else {
-		bcstEntries = mq.bcstWants.pending.SortedEntries()
+	mq.wllock.Lock()
+	peerEntries := mq.peerWants.pending.Entries()
+	bcstEntries := mq.bcstWants.pending.Entries()
+	cancels := mq.cancels.Keys()
+	if !supportsHave {
+		filteredPeerEntries := peerEntries[:0]
+		// If the remote peer doesn't support HAVE / DONT_HAVE messages,
+		// don't send want-haves (only send want-blocks)
+		//
+		// Doing this here under the lock makes everything else in this
+		// function simpler.
+		//
+		// TODO: We should _try_ to avoid recording these in the first
```

Comment:
I wonder if it would make sense to keep two lists:
- peerWants.wantHaves
- peerWants.wantBlocks

Reply:
That would simplify this.

Reply:
But later.